src/chainstate/stacks/index/storage.rs

```diff
@@ -1000,28 +1140,86 @@ impl TrieFileStorage {
-        
+
+        self.block_map.clear();
+        self.chain_tips.clear();
+
         Ok(())
     }
 
-    pub fn read_node_hash_bytes(&mut self, ptr: &TriePtr, buf: &mut Vec<u8>) -> Result<(), Error> {
+    /// Read a node's children's hashes as a vector of TrieHashes.
+    /// This only works for intermediate nodes and leafs (the latter of which have no children).
+    ///
+    /// This method is designed to only access hashes that are either (1) in this Trie, or (2) in
+    /// RAM already (i.e. as part of the block map)
+    ///
+    /// This means that the hash of a node that is in a previous Trie will _not_ be its
+    /// hash (as that would require a disk access), but would instead be the root hash of the Trie
+    /// that contains it.  While this makes the Merkle proof construction a bit more complicated,
+    /// it _significantly_ improves the performance of this method (which is crucial since this is on
+    /// the write path, which must be as short as possible).
+    ///
+    /// Rules:
+    /// If a node is empty, pass in an empty hash.
+    /// If a node is in this Trie, pass its hash.
+    /// If a node is in a previous Trie, pass the root hash of its Trie.
+    ///
+    /// On err, S may point to a prior block.  The caller should call s.open(...) if an error
+    /// occurs.
+    pub fn write_children_hashes<W: Write>(&mut self, node: &TrieNodeType, w: &mut W) -> Result<(), Error> {
+        trace!("get_children_hashes_bytes for {:?}", node);
+
+        let block_map = &self.block_map;
+
+        if Some(self.cur_block) == self.last_extended {
+            let hash_reader = self.last_extended_trie.as_mut()
+                .expect("MARF CORRUPTION: last_extended is set, but last_extended_trie is None.");
+            TrieFileStorage::inner_write_children_hashes(hash_reader, block_map, node, w)
+        } else {
+            let hash_reader = self.cur_block_fd.as_mut()
+                .ok_or(Error::NotFoundError)?;
+            TrieFileStorage::inner_write_children_hashes(hash_reader, block_map, node, w)
+        }
+    }
+
+    fn inner_write_children_hashes<W: Write, H: NodeHashReader>(
+        hash_reader: &mut H, block_map: &BlockHashMap, node: &TrieNodeType, w: &mut W) -> Result<(), Error> {
+        for ptr in node.ptrs().iter() {
+            if ptr.id() == TrieNodeID::Empty {
+                // hash of empty string
+                w.write_all(TrieHash::from_data(&[]).as_bytes())?;
+            }
+            else if !is_backptr(ptr.id()) {
+                // hash is in the same block as this node
+                hash_reader.write_node_hash(ptr, w)?;
+            }
+            else {
+                // AARON:
+                //   I *think* this is no longer necessary in the fork-table-less construction.
+                //   the back_pointer's consensus bytes uses this block_hash instead of a back_block
+                //   integer. This means that it would _always_ be included the node's hash computation.
```

Comment:
Let's test this -- can you add an `assert!()` here that checks this at runtime?

Reply:
I should clarify: this code path still gets visited, so I don’t know if an assert would be able to test anything here. But, I the byte array could be zeros here. If you look at the TriePtr.get_consensus_bytes method, you’ll see that it includes the block hash.